swim/join_delayer.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package swim
+
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/uber-common/bark"
+)
+
+var errAttemptNumberInvalid = errors.New("joinDelayer requires attemptNumber to be greater or equal to 0")
+var errLoggerRequired = errors.New("joinDelayer requires a logger")
+
+const (
+	defaultInitial = float64(100 * time.Millisecond)
+	defaultMax     = float64(60 * time.Second)
+)
+
+var delayerRand = rand.New(rand.NewSource(time.Now().UnixNano()))
+var defaultRandomizer = delayerRand.Intn
+var defaultSleeper = time.Sleep
+var noDelay = time.Duration(0)
+
+// joinDelayer defines the API of a delayer implementation that
+// applies a delay in between repeated join attempts.
+type joinDelayer interface {
+	delay(attemptNumber int) (time.Duration, error)
+}
+
+// delayRandomizer is a function that returns a random number between
+// 0 and an upper-bound, provided in its first argument.
+type delayRandomizer func(int) int
+
+// delaySleeper is a function that pauses execution for time.Duration.
+type delaySleeper func(time.Duration)
+
+// delayOpts is a struct that houses substitutable parameters of a joinDelayer
+// in order to alter its behavior.
+type delayOpts struct {
+	initial    float64
+	max        float64
+	randomizer delayRandomizer
+	sleeper    delaySleeper
+}
+
+// newDelayOpts creates a delayOpts struct with default values.
+func newDelayOpts() *delayOpts {
+	return &delayOpts{
+		initial:    defaultInitial,
+		max:        defaultMax,
+		randomizer: defaultRandomizer,
+		sleeper:    defaultSleeper,
+	}
+}
+
+// exponentialDelayer applies a delay in between repeated join attempts.
+// The delay increases exponentially and is capped at maxDelay.
+type exponentialDelayer struct {
+	// joiner is the address of the node sending join requests.
+	joiner string
```

Comment:
I am wondering if we could abstract everything join completely away from the delayer. It seems to me like the only thing we use join for is logging. We could make make this function emit events and do the logging in a handler that lives in the joiner. It doesn't seem like this is important right now but if we need different delayers I think we should make it more general.